packages/tinacms/src/components/ResizerBar.tsx

```diff
@@ -0,0 +1,129 @@
+/**
+
+Copyright 2019 Forestry.io Inc
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+
+import * as React from 'react'
+import styled from 'styled-components'
+import { radius, color } from '@tinacms/styles'
+import { SIDEBAR_MIN_WIDTH, Z_INDEX } from '../Globals'
+
+interface ResizerBarProps {
+  resizerBarWidth: number
+  open: boolean
+  sideBarWidth: number
+  setSideBarWidth: (nr: number) => void
+}
+
+export const ResizerBar = ({
+  resizerBarWidth,
+  open,
+  sideBarWidth,
+  setSideBarWidth,
+}: ResizerBarProps) => {
+  const [barPos, setBarPos] = React.useState(sideBarWidth - resizerBarWidth / 2)
+  const isResizing: any = React.useRef()
+
+  const resize = (e: any) => {
```

Comment:
It might be best to have `useCallback` wrap these:
* `resize`
* `handleMouseUp`
* `handleMouseDown`

Reply:
Is that necessary? I'm not using any useEffect in this one with the functions in a dependency array that will create an infinite loop. The recommendation is to use useCallback on expensive operations ... Some times the useCallback can actually decrease performance.

Reply:
Better safe then sorry. A new function instance is being created and passed to the `Resizer` component on every re-render. With the memo around it that won't happen, and we can be certain that nothing funky is happening. Every render may cause `Resizer` to re-render unnecessarily. I'm especially wary since those callbacks aren't just being used in React, they're being added and removed from the `document`.

Reply:
Ok. I can add them. This is an interesting discussion as there are many opinions on this. Some say that you don't need them and the useCallback itself create more overhead cost than leaving it out when it is small functions like this. =) The useCallback itself is a function that you have to call with a dependency array also ... 

Do you want to wrap the whole component in React.memo also or do you mean the useCallback on the three inline functions only?

Reply:
Pushed a new commit now. Added useCallback and also changed to mouse cursor to a resize cursor instead .. seems to be working fine. =)

Reply:
Took a more deep dive in thinking through this. I see why you want to have the useCallback on the callback functions to be safe. It's a good tool to use when certain child components shouldn't re-render when the parent component that holds the callback function re-renders. 

In this case the Styled Component named Resizer should actually re-render itself on each re-render. So the useCallback won't have any real effect on performance. I haven't measured anything but think it actually can slow stuff down as it is an extra function that does dependency checks via the dep. array (and also more code to be written). 

I'm not arguing against you. I'm just really curious cause we all have different approaches and opinions on stuff like this. =) It's a widely discussed topic in the React world and my learnings is that you don't need useCallback if it's not a real expensive function. Or if you don't want to re-render child components that gets the callback function as props. In this case the whole Resizer component is an "isolated" component that don't render any child components more than itself. 

And yet again ... I'm just curious to hear out your opinion cause you seem really experienced and I like discussions like this. It is you who has the final saying on the code and I will absolutely change it to however you want the code. Truly respect your knowledge!  =)

Reply:
I'm glad you did the dive @weibenfalk! I didn't dig too deeply on this particular case. 

I made the suggestion based on a heuristic: *if you're adding/removing the callbacks to something outside of the React Component it's safest to wrap them with `useCallback`*. In this situation, as well as with `Dismissible`, we're using the callbacks as event listeners on the document, so the heuristic applies.

In simpler examples I would opt for inline functions. In this counter example there's no need to memo-ize the callback:

```js
function Counter() {
  const [count, setCount] = useState(0)

  return <button onClick={() => setCount(count => count + 1)}> + </button>
} 
```

In that example there is little room for confusion or error. It might be mildly less performant but it's not a big deal. This component has callbacks, but those callbacks do not have any external side-effects.

*The main reason for this heuristic is the potential for memory leaks.* In the above example, a new `onClick` will be created on every render, but we know for certain that old one will be discarded. In the situations where were adding those callbacks to objects that live outside of our React component (i.e. the `document`) we need to be more careful. If we misunderstand how scope+hooks interact, then each render will create another function that never get garbage collected. This is what results in the memory leak. 

Using `useCallback` in those situations gives me some peace of mind. The potential risk of not using `useCallback` it is much higher than it's mental overhead cost.

Reply:
Thanks for that long answer. It's good to know how you're thinking. I also see your point as it is callbacks with references outside of the React Component. So .. in this case it was not for performance issues as I thought .. it is for have peace in mind. =) 

Regarding the inline function example. As it is a functional component it shouldn't matter if the function is inline. Both inlines and the other functions inside the functional component will get recreated on each re-render. So it should be acceptable to have it as a regular function also and not inline. =)

I've read a lot of opinions and also the React documents a few times since the release of hooks and they all say that useCallbacks isn't necessary for performance reasons. If you don't have a lot of child components that receive the callback as a prop. That will trigger re-renders of the child components also ... So the recommendation is to not overuse them cause you are "afraid" of performance issues with the function being recreated on each render. That's why I asked. 

Don't think I've come across any article about using them with callbacks that uses eventlisteners on document ... So your opinions is great reading for me! Love to learn more stuff!

Reply:
By "inline" I meant "not-memoized". 

The above example is equivalent to:

```js
function Counter() {
  const [count, setCount] = useState(0)

  const increment = () => setCount(count => count + 1)

  return <button onClick={increment}> + </button>
}
```

Again, I haven't dug into this. It's quite possibly fine. I would personally just use `useCallback` because I'm certain it will be fine. Convincing myself that not using it is fine would take more work then I'm willing to put in. It's a decision based on expediency rather than correctness. 

If after digging in you find that this does not endlessly add callbacks to the document and does not trigger expensive re-renders, then not wrapping them is cool with me.

Reply:
No no .. it's totally fine! I was just curious and also fun to discuss React topics as I'm passionate about React.